illumos-utils/src/running_zone.rs

```diff
@@ -586,8 +727,7 @@ impl RunningZone {
     ///
     /// Allows callers to synchronously stop a zone, and inspect an error.
     pub async fn stop(&mut self) -> Result<(), String> {
```

Comment:
Could this take `mut self` instead of `&mut self`? With the change from `self.running` to `self.id.take()`, if a caller incorrectly calls `run_cmd()` after `stop()` it will now panic on the `self.id.expect(...)` in `run_cmd()`. Taking `mut self` here would make that impossible, although it would also mean a failure here would leave the zone in a zombie state.

Maybe the zombie state issue is already present regardless of `&mut self` though - if `halt_and_remove_logged()` fails, we've already stolen `self.id`, which means neither another `stop()` nor `drop()` will try again to shut down the zone. It's unclear to me what a caller would expect if `stop()` fails, though.

Reply:
I agree, this isn't great. At least we should check for `self.id.is_some()` in `run_cmd()`, to avoid the panic. It's probably a programming error to call `stop()` more than once, but we should still not panic I think.

I think we can't take self by value here, at least not without more significant changes at the call sites. The zombie issue was already there, even with the boolean. Calling `stop()` only works at most once, since it previously set `self.running = false;`. IIRC, the point of `stop()` was just to be able to inspect / log the failure.